src/infql.jl

```diff
@@ -378,3 +378,216 @@ function LazyBandedMatrices._SymTridiagonal(::Tuple{TriangularLayout{'L', 'N', P
+    j = maximum(inds)
+    maxN = 1000*j # Prevent runaway loop
+    Lerr = one(T)
+    N = j
+    checkinds = max(1,j-bandwidth(A.M,1)-bandwidth(A.M,2))
+    @inbounds Ls = ql(A.M[checkinds:N,checkinds:N]).L[2:j-checkinds+1,2:j-checkinds+1]
+    @inbounds while Lerr > A.tol
+        # compute QL for small finite section and large finite section
+        Ll = ql(A.M[checkinds:2N,checkinds:2N]).L[2:j-checkinds+1,2:j-checkinds+1]
+        # compare bottom right sections and stop if desired level of convergence achieved
+        Lerr = norm(Ll-Ls,2)
+        if N == maxN
+            error("Reached max. iterations in finite section QL without convergence to desired tolerance.")
+        end
+        Ls = Ll
+        N = 2*N
+    end
+    A.data = ql(A.M[1:(N÷2),1:(N÷2)]).factors[1:j,1:j]
+end
+
+
+function cache_filldata!(A::AdaptiveQLTau{T}, inds::UnitRange{Int}) where T
+    j = maximum(inds)
+    maxN = 1000*j
+    Lerr = one(T)
+    N = j
+    checkinds = max(1,j-bandwidth(A.M,1)-bandwidth(A.M,2))
+    @inbounds Ls = ql(A.M[checkinds:N,checkinds:N]).L[2:j-checkinds+1,2:j-checkinds+1]
+    @inbounds while Lerr > A.tol
+        # compute QL for small finite section and large finite section
+        Ll = ql(A.M[checkinds:2N,checkinds:2N]).L[2:j-checkinds+1,2:j-checkinds+1]
+        # compare bottom right sections and stop if desired level of convergence achieved
+        Lerr = norm(Ll-Ls,2)
+        if N == maxN
+            error("Reached max. iterations in finite section QL without convergence to desired tolerance.")
+        end
+        Ls = Ll
+        N = 2*N
+    end
+    A.data = ql(A.M[1:(N÷2),1:(N÷2)]).τ[1:j]
+end
+
+function getindex(K::AdaptiveQLFactors, k::Integer, j::Integer)
+    resizedata!(K, max(k,j))
+    K.data[k, j]
+end
+function getindex(K::AdaptiveQLFactors, k::Integer, jr::UnitRange{Int})
+    resizedata!(K, max(k,maximum(jr)))
+    K.data[k, jr]
+end
+function getindex(K::AdaptiveQLFactors, kr::UnitRange{Int}, j::Integer)
+    resizedata!(K, max(j,maximum(kr)))
+    K.data[kr, j]
+end
+function getindex(K::AdaptiveQLFactors, kr::UnitRange{Int}, jr::UnitRange{Int})
+    resizedata!(K, max(maximum(jr),maximum(kr)))
+    K.data[kr, jr]
+end
+function getindex(K::AdaptiveQLTau, jr::UnitRange{Int})
+    resizedata!(K, maximum(jr))
+    K.data[jr]
+end
+function getindex(K::AdaptiveQLTau, k::Integer)
+    resizedata!(K, k)
+    K.data[k]
+end
+
+function getproperty(F::QL{T, AdaptiveQLFactors{T}, AdaptiveQLTau{T}}, d::Symbol) where T
+    if d == :L
+        return LowerTriangular(F.factors)
+    elseif d == :Q
+        return getQ(F)
+    else
+        getfield(F, d)
+    end
+end
+
+@inline function getQ(F::QL{T, AdaptiveQLFactors{T}, AdaptiveQLTau{T}}) where T
+    Q = AdaptiveQLQ{T}([one(T)],F.factors,F.τ,(0,0))
+    # initialize as 50x50, then adaptively expand
+    resizedata!(Q,50)
+    return Q
+end
+
+mutable struct AdaptiveQLQ{T} <: AbstractCachedMatrix{T}
+    data
```

Comment:
Shouldn't this be strongly typed? (eg `::Matrix{T}`)

Reply:
yes, data field can be set to `::Matrix{T}`